### ISSUE TITLE
wrong-type-argument eieio-object-p nil

### ISSUE THREAD

**Issue by synic** [syl20bnr/spacemacs#5020]
I keep getting this after updating packages.  Removing elpa fixes it, but that would only be ok if it happened once or twice.  It happens often.

What gives?

#### System Info
- OS: darwin
- Emacs: 24.5.7
- Spacemacs: 0.105.9
- Spacemacs branch: develop (rev. de7eb40)
- Distribution: spacemacs
- Editing style: vim
- Completion: helm
- Layers:
```elisp
(unscroll auto-completion xkcd emacs-lisp git dash yaml github javascript html markdown extra-langs python ruby org osx python django syntax-checking spell-checking version-control lua colors spacemacs-layouts vimscript spacemacs-helm
          (c-c++ :variables c-c++-enable-clang-support t)
          (shell :variables shell-default-height 30 shell-default-position 'bottom))
```

**Comment by nwolfe** [comment]
Yeah I've been dealing with the same thing.  It's often the `spinner` package that produces the error message about `wrong-type-argument eieio-object-p nil`.  I tried looking for a more targeted fix other than blowing away elpa/ but didn't have any luck.

**Comment by travisbhartwell** [comment]
I can verify the same thing, but on Linux.  I, too, resolved it by wiping my elpa directory.

**Comment by StreakyCobra** [comment]
I have the same, no idea where it come from.

**Comment by nixmaniack** [comment]
And I can confirm too but without much debugging. Just deleted elpa.

**Comment by bmag** [comment]
Encountered the same just now, for `helm-core`, `spinner` and `packed`, but it looks like the packages are installed successfully. I am able to reproduce the error by deleting the `packed-*` directory inside `elpa` and restarting emacs. Will debug further.

**Comment by bmag** [comment]
From [`configuration-layer//install-package`](https://github.com/syl20bnr/spacemacs/blob/a111f4802d850c32a2ef6ff6c2d62cc6c0861b84/core/core-configuration-layer.el#L713):
```elisp
          (cond
           ((or (null pkg) (eq 'elpa location))
            (configuration-layer//install-from-elpa pkg-name)
            (oset pkg :lazy-install nil))
           ((and (listp location) (eq 'recipe (car location)))
            (configuration-layer//install-from-recipe pkg)
            (oset pkg :lazy-install nil))
           (t (spacemacs-buffer/warning "Cannot install package %S."
                                        pkg-name)))
```

The `(or (null pkg) (eq 'elpa location))` branch calls `(oset pkg :lazy-install nil)` even if `pkg` is nil, and that throws a `wrong-type-argument` error.
The error is thrown *after* the package is really installed, so *you can just ignore it* for now, until the bug is fixed. No need to delete elpa.

**Comment by sooheon** [comment]
Getting this while calling `ivy-spacemacs-help`, `ivy-spacemacs-help-layers`, `ivy-spacemacs-help-docs` on latest develop after nuked elpa. @justbur could you weigh in on what it is in those functions that could be causing this?

**Comment by justbur** [comment]
I haven't seen this error. `ivy-spacemacs-help//init` uses functions in core-configuration-layer.el to build a list of layers and packages the same way the corresponding helm does. It seems there are some bugs in the core-configuration-layer functions

It's potentially related to a specific layer, possibly a private one. A backtrace would really help here.

**Comment by sooheon** [comment]
```
Debugger entered--Lisp error: (wrong-type-argument eieio-object-p nil)
  signal(wrong-type-argument (eieio-object-p nil))
  eieio-oset(nil :lazy-install t)
  (let ((layer (object-assoc layer-name :name configuration-layer--layers)) (extensions (spacemacs/mplist-get props :extensions))) (eieio-oset layer (quote :lazy-install) t) (let ((--dolist-tail-- extensions) x) (while --dolist-tail-- (setq x (car --dolist-tail--)) (let ((ext (car x)) (mode (cadr x))) (add-to-list (quote configuration-layer--lazy-mode-alist) (cons mode ext)) (add-to-list (quote auto-mode-alist) (list ext (quote lambda) nil (list (quote configuration-layer//auto-mode) (list ... layer-name) (list ... mode))))) (setq --dolist-tail-- (cdr --dolist-tail--)))))
  (progn (let ((layer (object-assoc layer-name :name configuration-layer--layers)) (extensions (spacemacs/mplist-get props :extensions))) (eieio-oset layer (quote :lazy-install) t) (let ((--dolist-tail-- extensions) x) (while --dolist-tail-- (setq x (car --dolist-tail--)) (let ((ext (car x)) (mode (cadr x))) (add-to-list (quote configuration-layer--lazy-mode-alist) (cons mode ext)) (add-to-list (quote auto-mode-alist) (list ext (quote lambda) nil (list ... ... ...)))) (setq --dolist-tail-- (cdr --dolist-tail--))))))
  (if dotspacemacs-enable-lazy-installation (progn (let ((layer (object-assoc layer-name :name configuration-layer--layers)) (extensions (spacemacs/mplist-get props :extensions))) (eieio-oset layer (quote :lazy-install) t) (let ((--dolist-tail-- extensions) x) (while --dolist-tail-- (setq x (car --dolist-tail--)) (let ((ext ...) (mode ...)) (add-to-list (quote configuration-layer--lazy-mode-alist) (cons mode ext)) (add-to-list (quote auto-mode-alist) (list ext ... nil ...))) (setq --dolist-tail-- (cdr --dolist-tail--)))))))
  configuration-layer/lazy-install(elixir :extensions ("\\.\\(ex\\|exs|elixir\\)\\'" elixir-mode))
  eval-buffer(#<buffer  *load*> nil "/Users/sooheon/.emacs.d/layers/+lang/elixir/config.el" nil t)  ; Reading at buffer position 407
  load-with-code-conversion("/Users/sooheon/.emacs.d/layers/+lang/elixir/config.el" "/Users/sooheon/.emacs.d/layers/+lang/elixir/config.el" nil nil)
  load("/Users/sooheon/.emacs.d/layers/+lang/elixir/config.el")
  (if (file-exists-p file) (load file))
  (let ((file (concat (eieio-oref layer (quote :dir)) file))) (if (file-exists-p file) (load file)))
  (while --dolist-tail-- (setq file (car --dolist-tail--)) (let ((file (concat (eieio-oref layer (quote :dir)) file))) (if (file-exists-p file) (load file))) (setq --dolist-tail-- (cdr --dolist-tail--)))
  (let ((--dolist-tail-- files) file) (while --dolist-tail-- (setq file (car --dolist-tail--)) (let ((file (concat (eieio-oref layer (quote :dir)) file))) (if (file-exists-p file) (load file))) (setq --dolist-tail-- (cdr --dolist-tail--))))
  configuration-layer//load-layer-files([object cfgl-layer "elixir" elixir "/Users/sooheon/.emacs.d/layers/+lang/elixir/" nil nil nil nil] ("funcs.el" "config.el"))
  (if (configuration-layer/layer-usedp (eieio-oref layer (quote :name))) nil (configuration-layer//load-layer-files layer (quote ("funcs.el" "config.el"))))
  (while --dolist-tail-- (setq layer (car --dolist-tail--)) (if (configuration-layer/layer-usedp (eieio-oref layer (quote :name))) nil (configuration-layer//load-layer-files layer (quote ("funcs.el" "config.el")))) (setq --dolist-tail-- (cdr --dolist-tail--)))
  (let ((--dolist-tail-- ivy-spacemacs-help-all-layers) layer) (while --dolist-tail-- (setq layer (car --dolist-tail--)) (if (configuration-layer/layer-usedp (eieio-oref layer (quote :name))) nil (configuration-layer//load-layer-files layer (quote ("funcs.el" "config.el")))) (setq --dolist-tail-- (cdr --dolist-tail--))))
  (progn (mapc (function (lambda (layer) (setq ivy-spacemacs-help-all-layers (cons (configuration-layer/make-layer layer) ivy-spacemacs-help-all-layers)))) (configuration-layer/get-layers-list)) (let ((--dolist-tail-- ivy-spacemacs-help-all-layers) layer) (while --dolist-tail-- (setq layer (car --dolist-tail--)) (if (configuration-layer/layer-usedp (eieio-oref layer (quote :name))) nil (configuration-layer//load-layer-files layer (quote ("funcs.el" "config.el")))) (setq --dolist-tail-- (cdr --dolist-tail--)))) (setq ivy-spacemacs-help-all-packages (configuration-layer/get-packages ivy-spacemacs-help-all-layers)))
  (if (or arg (null ivy-spacemacs-help-all-packages)) (progn (mapc (function (lambda (layer) (setq ivy-spacemacs-help-all-layers (cons (configuration-layer/make-layer layer) ivy-spacemacs-help-all-layers)))) (configuration-layer/get-layers-list)) (let ((--dolist-tail-- ivy-spacemacs-help-all-layers) layer) (while --dolist-tail-- (setq layer (car --dolist-tail--)) (if (configuration-layer/layer-usedp (eieio-oref layer (quote :name))) nil (configuration-layer//load-layer-files layer (quote ("funcs.el" "config.el")))) (setq --dolist-tail-- (cdr --dolist-tail--)))) (setq ivy-spacemacs-help-all-packages (configuration-layer/get-packages ivy-spacemacs-help-all-layers))))
  ivy-spacemacs-help//init()
  ivy-spacemacs-help()
  call-interactively(ivy-spacemacs-help nil nil)
  command-execute(ivy-spacemacs-help)
```

Here's the backtrace I got.

**Comment by justbur** [comment]
Same thing that @bmag found

**Comment by sooheon** [comment]
Yep, should have read more carefully. Thanks :)

**Comment by bmag** [comment]
Same error signal, but a different bug than the one I fixed. In [this line](https://github.com/syl20bnr/spacemacs/blob/e061b13eb5a37c5331b75da87fa82ae01032a31a/core/core-configuration-layer.el#L641) of `configuration-layer/lazy-install`, we fail to set the `:lazy-install` attribute of `layer`, because `layer` is nil. @sooheon can you open a new issue so we can track it there?

It seems that the call `(object-assoc 'elixir :name configuration-layer--layers)` returned nil, where it should have returned elixir's layer (a `cfgl-layer` object). Meaning that elixir is missing from `configuration-layer--layers`. @sooheon please include the output of `SPC h d s`, so we can see what layers you enabled.

**Comment by geraldodev** [comment]
I'm getting this from fresh spacemacs install.
(Bootstrap) Installing evil...          
(Bootstrap) Installing bind-map...
(Bootstrap) Installing bind-key... 
(Bootstrap) Installing which-key...
(Bootstrap) Installing use-package...                                                                    
(Bootstrap) Installing package-build...                                                                  
(Bootstrap) Installing quelpa...          

It's a show stopper because it does not get to the point that evil is loaded
Restarting does not help

GNU Emacs 24.5.1
Linux 4.4.5-1-ARCH #1 SMP PREEMPT Thu Mar 10 07:38:19 CET 2016 x86_64 GNU/Linux

```
Debugger entered--Lisp error: (wrong-type-argument (or eieio-object-p class-p) nil)
  signal(wrong-type-argument ((or eieio-object-p class-p) nil))
  eieio-oref(nil :variables)
  (let ((variables (eieio-oref layer (quote :variables)))) (while variables (let ((var (car-safe (prog1 $
  configuration-layer//set-layer-variables(nil)
  configuration-layer//configure-layer(nil)
  (while --dolist-tail-- (setq l (car --dolist-tail--)) (configuration-layer//configure-layer l) (setq -$
  (let ((--dolist-tail-- layers) l) (while --dolist-tail-- (setq l (car --dolist-tail--)) (configuration$
  (let ((warning-minimum-level :error)) (let ((--dolist-tail-- layers) l) (while --dolist-tail-- (setq l$
  configuration-layer//configure-layers((nil [object cfgl-layer "emacs-lisp" emacs-lisp "/home/geraldo/.$
  configuration-layer/sync()
  (if (not (version<= spacemacs-emacs-min-version emacs-version)) (message (concat "Your version of Emac$
  eval-buffer(#<buffer  *load*> nil "/home/geraldo/.emacs.d/init.el" nil t)  ; Reading at buffer positio$
  load-with-code-conversion("/home/geraldo/.emacs.d/init.el" "/home/geraldo/.emacs.d/init.el" t t)
  load("/home/geraldo/.emacs.d/init" t t)
  #[0 "^H\205\262^@     \306=\203^Q^@\307^H\310Q\202;^@ \311=\204^^^@\307^H\312Q\202;^@\313\307\314\315#$
  command-line()
  normal-top-level()
```

**Comment by bmag** [comment]
@geraldodev It's the same error signal, but probably a different bug. Can you please:
- open a new issue
- include the output of `M-x spacemacs/describe-system-info`
- repost the error, but make sure no lines are truncated (what you posted contains some lines that were truncated - these are the lines that end in `$`)

**Comment by whyname** [comment]
the same with @geraldodev found, if this fixed?

**Comment by d12frosted** [comment]
Fixed with released of Spacemacs v0.200. Let us know if you still have any problems with this issue.
